src/components/VmActions/AutoSelectConsole.js

```diff
@@ -0,0 +1,133 @@
+import React, { useEffect, useState } from 'react'
+import { connect } from 'react-redux'
+import PropTypes from 'prop-types'
+import { addUserMessage, downloadVmConsoles } from '_/actions'
+import ConsoleConfirmationModal from '_/components/VmActions/ConsoleConfirmationModal'
+import { msg } from '_/intl'
+
+const AutoSelectConsole = (props) => {
+  const {
+    vm,
+    show,
+    defaultConsole,
+    isNoVNC,
+    fetchConsoles,
+    requestActive,
+    id,
+    onFailure,
+  } = props
+
+  const initialState = {
+    // consoles: [],
+    selectedProtocol: undefined,
+    failureReason: undefined,
+    dispatchStatus: 'doNotDispatch',
+    active: false,
+  }
+
+  const [state, setState] = useState(initialState)
+  const {
+    dispatchStatus,
+    selectedProtocol,
+    active,
+    // failureReason,
+    // consoles,
+  } = state
+
+  useEffect(() => {
+    if (show) {
+      setState(state => ({
+        ...state,
+        dispatchStatus: 'shouldDispatch',
+        active: true,
+      }))
+      fetchConsoles(vm.get('id'))
+    }
+  }, [show]) // eslint-disable-line react-hooks/exhaustive-deps
+
+  useEffect(() => {
+    const newState = {}
+
+    const { selectedProtocol, failureReason } = selectProtocol(vm.get('consoles'), requestActive)
+
+    if (selectedProtocol || failureReason) {
+      newState.selectedProtocol = selectedProtocol
+      newState.failureReason = failureReason
+    }
+    if (active && selectedProtocol && dispatchStatus !== 'doNotDispatch') newState.dispatchStatus = 'dispatched' // cannot change from 'doNotDispatch' to 'dispatched'
+    if (active && !selectedProtocol && failureReason && !requestActive && dispatchStatus !== 'doNotDispatch') onFailure(failureReason)
+
+    setState(state => ({ ...state, ...newState }))
+  }, [vm, requestActive]) // eslint-disable-line react-hooks/exhaustive-deps
+
+  const onClose = () => {
```

Comment:
why we need `onClose` here?

Reply:
To update the state before call the `props.onClose` function, but it's changed so take a look on the latest version :)